### PR TITLE
feat: Add new distill losses

### PR DESCRIPTION
**Pull request by HannanJaved**
#### Reference Issues/PRs
Issue #238 - Incorporate knowledge distillation losses

#### What does this implement/fix? Explain your changes.
Added L1, L2, reverse_kld, cosine similarity, and Jensen-Shannon Divergence losses as options for distillation loss

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the
terms of your choice.

### CODE REVIEW

File: test/test_training_strategies.py [hunk 1]
```diff
@@ -80,7 +80,7 @@ def test_integration_training_strategies_mlp(strategy):
 
 
 @pytest.mark.parametrize("strategy", methods)
-@pytest.mark.parametrize("kd_loss", [None, DistillLoss(0.5, 0.5)])
+@pytest.mark.parametrize("kd_loss", [None, DistillLoss(0.5, 0.5, "kld")])
```
Comment (aaronkl) [inline test/test_training_strategies.py:83]: Could you also add a unit test for all the other loss functions?

File: whittle/loss/kd_loss.py [hunk 2]
```diff
@@ -19,19 +19,22 @@ class DistillLoss(nn.Module):
         kldiv (nn.KLDivLoss): The KL divergence loss function.
     """
 
-    def __init__(self, temperature, distillation_weight):
+    def __init__(self, temperature, distillation_weight, loss):
```
Comment (aaronkl) [inline whittle/loss/kd_loss.py:22]: Could we add types here? Could be also in a separate PR

File: whittle/loss/kd_loss.py [hunk 3]
```diff
@@ -19,19 +19,22 @@ class DistillLoss(nn.Module):
         kldiv (nn.KLDivLoss): The KL divergence loss function.
```
Comment (aaronkl) [inline whittle/loss/kd_loss.py:19]: Could you update the docsting?

**Review by rheasukthanker — CHANGES_REQUESTED** [review]
Ideally this PR should include all the KD losses we have studied in previous work. Refer to https://github.com/automl/automated-sub-network-selection/blob/main/whittle/loss/kd_losses.py and https://github.com/automl/automated-sub-network-selection/blob/main/whittle/loss/loss_factory.py. Could we if possible include all different loss variants in this PR?

File: whittle/loss/kd_loss.py [hunk 4]
```diff
@@ -53,10 +53,48 @@ def forward(self, outputs, labels, outputs_teacher) -> nn.Tensor:
         outputs_teacher = outputs_teacher.detach()
 
         if outputs_teacher is not None and self.distillation_weight > 0:
-            soft_target_loss = self.kldiv(
-                F.log_softmax(outputs / self.temperature, dim=1),
-                F.softmax(outputs_teacher / self.temperature, dim=1),
-            ) * (self.temperature**2)
+            if self.distill_loss == "kld":
```
Comment (rheasukthanker) [inline whittle/loss/kd_loss.py:56]: Could we rename the losses to "forward_kld" "reverse_kld" to be more consistent with literature?

File: whittle/loss/kd_loss.py [hunk 5]
```diff
@@ -53,10 +53,48 @@ def forward(self, outputs, labels, outputs_teacher) -> nn.Tensor:
...
+
+            elif self.distill_loss == "l2" or self.distill_loss == "mse":
+                soft_target_loss = F.mse_loss(outputs, outputs_teacher, reduction="mean")
+
+            elif self.distill_loss == "cosine":
+                soft_target_loss = (
+                    1 - F.cosine_similarity(outputs, outputs_teacher, dim=1).mean()
+                )
+
+            elif self.distill_loss == "jsd":
```
Comment (rheasukthanker) [inline whittle/loss/kd_loss.py:81]: Rename to "js_distance"

File: whittle/loss/kd_loss.py [hunk 6]
```diff
@@ -53,10 +53,48 @@ def forward(self, outputs, labels, outputs_teacher) -> nn.Tensor:
...
+                        F.log_softmax(outputs_teacher, dim=1),
+                        F.softmax(m, dim=1),
+                        reduction="batchmean",
+                    )
+                )
+
+            else:
+                raise ValueError(f"Invalid distillation loss: {self.distill_loss}")
 
         hard_target_loss = F.cross_entropy(outputs, labels, reduction="mean")
```
Comment (rheasukthanker) [inline whittle/loss/kd_loss.py:100]: For total loss, could we also allow for different weights for each loss factor? Specifically they need not sum up to one and it might make sense to have custom weighing factors for each term.

File: whittle/loss/kd_loss.py (hunk 4 above)
Comment (rheasukthanker) [reply whittle/loss/kd_loss.py:56]: Minor comment: I think it is much cleaner to move each loss type/equation to individual methods that can be called to make the sequence of if..elif statements a bit cleaner. This is also better in the long term if we add newer losses.

File: test/test_distill_loss.py [hunk 7]
```diff
@@ -11,9 +11,18 @@ def test_distill_loss():
...
+    beta = 0.5
+    loss_type = "forward_kld"
+    weight_scheme = "other"
+
+    loss_fn = DistillLoss(
+        alpha=alpha,
+        beta=beta,
+        temperature=temperature,
+        loss=loss_type,
+        weight_scheme=weight_scheme,
```
Comment (rheasukthanker) [inline test/test_distill_loss.py:24]: @HannanJaved  Could we remove weight scheme argument from the distill loss? I didn't end up using it for our experiments and it is best not to add it unless we see it is empirically useful.

File: whittle/args.py [hunk 8]
```diff
@@ -26,13 +26,18 @@ class DistillArgs:
...
+        weight_scheme: Weight scheme to use for the distillation loss. Options are 'default' (use alpha=1 and beta=hard_target_loss/soft_target_loss).
     """
 
     method: str = "logits"
-    temperature: float = 5
     alpha: float = 0.5
+    beta: float = 0.5
+    temperature: float = 5
+    loss: str = "forward_kld"
+    weight_scheme: str = "default"
```
Comment (rheasukthanker) [inline whittle/args.py:40]: Could we set weight_scheme arg that the args falls back to to "other"? Empirically weight_scheme: str = "default" didn't work well for inplace kd.

**Review by rheasukthanker — APPROVED** [review]
Looks good and ready to be merged (after addressing a minor comment)

File: whittle/args.py (hunk 8 above)
Comment (aaronkl) [reply whittle/args.py:40]: Actually what do you think about omitting this parameter and just change the default values for alpha=1 and beta=None. In DistillLoss we could then do `beta = hard_target_loss/soft_target_loss if beta is None`. wdyt?

Comment (aaronkl) [reply whittle/args.py:40]: could be also in a separate PR

Comment (HannanJaved) [reply whittle/args.py]: I have updated the code so that weight_scheme is set to "other", alpha as 1, and beta as 0 by default.
Should I merge it? If there are other changes necessary, maybe we can do it in another PR?
